main/ruletester/ruletester.go

```diff
@@ -72,132 +105,295 @@ func main() {
+	metrics, err := ReadMetricsFile(*metricsFile)
+	if err != nil {
+		fmt.Printf("Fatal error reading metrics file %s\n", err)
+		os.Exit(1)
+	}
+
+	DoAnalysis(metrics, graphiteConverter)
+
+	fmt.Printf("Total metric count %d\n", len(metrics))
+
+	// matched := 0
+	// unmatched := 0
+	// reverse_convert_failed := 0
+
+	// fmt.Printf("Matched: %d\n", matched)
+	// fmt.Printf("Unmatched: %d\n", unmatched)
+	// fmt.Printf("Reverse convert failed: %d\n", reverse_convert_failed)
 
-	metricFile, err := os.Open(*metricsFile)
 	if err != nil {
 		common.ExitWithMessage("No metric file.")
 	}
-	scanner := bufio.NewScanner(metricFile)
+	// scanner := bufio.NewScanner(metricFile)
 	cassandraConfig := cassandra.CassandraMetricMetadataConfig{
 		Hosts:    config.MetricMetadataConfig.Hosts,
 		Keyspace: config.MetricMetadataConfig.Keyspace,
 	}
-	apiInstance := common.NewMetricMetadataAPI(cassandraConfig)
+	_ = common.NewMetricMetadataAPI(cassandraConfig)
 
-	var output *os.File
-	if *unmatchedFile != "" {
-		output, err = os.Create(*unmatchedFile)
-		if err != nil {
-			common.ExitWithMessage(fmt.Sprintf("Error creating the output file: %s", err.Error()))
-		}
-	}
-	stat := run(ruleset, scanner, apiInstance, output)
-	report(stat)
+	// var output *os.File
+	// if *unmatchedFile != "" {
+	// 	output, err = os.Create(*unmatchedFile)
+	// 	if err != nil {
+	// 		common.ExitWithMessage(fmt.Sprintf("Error creating the output file: %s", err.Error()))
+	// 	}
+	// }
+	// stat := run(ruleset, scanner, apiInstance, output)
+	// report(stat)
 }
 
-func run(ruleset util.RuleSet, scanner *bufio.Scanner, apiInstance api.MetricMetadataAPI, unmatched *os.File) Statistics {
+type ChunkResult struct {
+	matched                int
+	unmatched              int
+	reverse_convert_failed int
+}
+
+func DoAnalysis(metrics []string, graphiteConverter util.RuleBasedGraphiteConverter) {
+	graphiteConverter.EnableStats()
+
+	workQueue := make(chan []string, len(metrics)/25+1)
+	resultQueue := make(chan ChunkResult, len(metrics))
+	unmatchedQueue := make(chan string, len(metrics))
 	var wg sync.WaitGroup
-	stat := Statistics{
-		perMetric: make(map[api.MetricKey]PerMetricStatistics),
+
+	i := 0
+	for i = 0; i+25 < len(metrics); i = i + 25 {
+		workQueue <- metrics[i : i+25]
 	}
-	type result struct {
-		input   string
-		result  api.TaggedMetric
-		success bool
+	if i < len(metrics) {
+		workQueue <- metrics[i:]
 	}
-	inputBuffer := make(chan string, 10)
-	outputBuffer := make(chan result, 10)
-	done := make(chan struct{})
-	for id := 0; id < runtime.NumCPU(); id++ {
+
+	close(workQueue)
+
+	fmt.Printf("Starting work...\n")
+	for j := 0; j < 10; j++ {
+		wg.Add(1)
 		go func() {
+			counter := 0
+			defer wg.Done()
 			for {
 				select {
-				case <-done:
-					return
-				case input := <-inputBuffer:
-					metric, matched := ruleset.MatchRule(input)
-					outputBuffer <- result{
-						input,
-						metric,
-						matched,
+				case metrics, more := <-workQueue:
```

Comment:
The `for { select {` can just be changed into a range over the channel.

`for metrics := range workQueue {`